core/services/route_match/RouteMatchSpecificationFactory.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace EventEspresso\core\services\route_match;
+
+use EventEspresso\core\domain\entities\route_match\RouteMatchSpecification;
+use EventEspresso\core\exceptions\InvalidDataTypeException;
+use EventEspresso\core\exceptions\InvalidInterfaceException;
+use EventEspresso\core\interfaces\FactoryInterface;
+use EventEspresso\core\services\loaders\LoaderFactory;
+use EventEspresso\core\services\loaders\LoaderInterface;
+use InvalidArgumentException;
+use ReflectionException;
+
+/**
+ * Class RouteMatchSpecificationFactory
+ * Factory for generating RouteMatchSpecification classes.
+ * Will automatically resolve dependencies using the
+ * RouteMatchSpecificationDependencyResolver
+ *
+ * @package EventEspresso\core\services\route_match
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+class RouteMatchSpecificationFactory implements FactoryInterface
+{
+
+    /**
+     * @var RouteMatchSpecificationDependencyResolver $dependency_resolver
+     */
+    private $dependency_resolver;
+
+    /**
+     * @var LoaderInterface $loader
+     */
+    private $loader;
+
+
+    /**
+     * RouteMatchSpecificationFactory constructor
+     *
+     * @param RouteMatchSpecificationDependencyResolver $dependency_resolver
+     * @param LoaderInterface                           $loader
+     */
+    public function __construct(RouteMatchSpecificationDependencyResolver $dependency_resolver, LoaderInterface $loader)
```

Comment:
I realize its still early and you might not want to act on my comment yet because of that, but this almost appears like a potential pattern that could occur for other factories that require a resolver.  So my thinking is there may be good use case for something like:

```php
abstract class FactoryWithResolver implements FactoryInterface {
    private $dependency_resolver;
    private $loader;
    public function __construct(DependencyResolverInterface $dependency_resolver, LoaderInterface $loader) {
        $this->dependency_resolver = $dependency_resolver;
        $this->loader = $loader;
    }

    public function createInstance($fqcn) {
        $this->dependency_resolver->resolveDependencies($fqcn);
        return $this->loader->getShared($fqcn);
    }
}
```

Child classes would still need to implement the static `create` method, but this could eliminate repeated boilerplate.  

Again, I know there's no demonstrated need for this additional abstraction yet so understood if you decide to ignore this comment for now.

Reply:
ya I think that's a great idea. I think there's a few types of classes (like `EE_DMS_Core_*` or shortcodes) that have identical dependencies hardcoded in `EE_Dependency_Map` that could benefit from this. We could likely even abstract the `RouteMatchSpecificationDependencyResolver` out into a parent `DependencyResolver` class as well, since all such classes would likely require the same dependencies, and likely behave in a very similar fashion.